File: Core/GameEngine.cpp
#include "pch.h"
#include "GameEngine.h"

#include <algorithm>
#include <functional>
#include <optional>
#include <random>

#include <winrt/Microsoft.UI.Dispatching.h>

namespace how_to_train_your_nailong::Game
{
    namespace
    {
        // Hard-coded fallback difficulty presets. Mirror the values in
        // Assets/Config/game_difficulty.json so behaviour is consistent even if
        // that file is missing or corrupt. A future change can have the engine
        // load and merge the JSON over these defaults.
        DifficultyParams ParamsFor(Difficulty d) noexcept
        {
            DifficultyParams p;
            switch (d)
            {
            case Difficulty::Easy:
                p.nailong_laugh_min = Duration{8000};
                p.nailong_laugh_max = Duration{18000};
                p.calibration       = Duration{3000};
                p.smile_threshold   = 0.30f;
                p.consecutive_frames_to_confirm = 4;
                p.lost_face_grace   = Duration{2000};
                p.lost_face_invalid = Duration{7000};
                break;
            case Difficulty::Normal:
                p.nailong_laugh_min = Duration{5000};
                p.nailong_laugh_max = Duration{12000};
                p.calibration       = Duration{2500};
                p.smile_threshold   = 0.28f;
                p.consecutive_frames_to_confirm = 4;
                p.lost_face_grace   = Duration{1500};
                p.lost_face_invalid = Duration{5000};
                break;
            case Difficulty::Hard:
                p.nailong_laugh_min = Duration{3000};
                p.nailong_laugh_max = Duration{8000};
                p.calibration       = Duration{2000};
                p.smile_threshold   = 0.25f;
                p.consecutive_frames_to_confirm = 3;
                p.lost_face_grace   = Duration{1000};
                p.lost_face_invalid = Duration{4000};
                break;
            }
            return p;
        }
    }

    struct GameEngine::Impl
    {
        IGameView&        view;
        GameState         state{GameState::Idle};
        Difficulty        difficulty{Difficulty::Normal};
        DifficultyParams  params{ParamsFor(Difficulty::Normal)};
        Winner            last_winner{Winner::None};

        std::mt19937                       rng{std::random_device{}()};
        TimePoint                          nailong_laugh_deadline{};
        std::optional<TimePoint>           lost_face_since;
        int                                countdown_remaining{0};

        winrt::Microsoft::UI::Dispatching::DispatcherQueue       dispatcher{nullptr};
        winrt::Microsoft::UI::Dispatching::DispatcherQueueTimer  delay_timer{nullptr};
        winrt::Microsoft::UI::Dispatching::DispatcherQueueTimer  countdown_timer{nullptr};
        winrt::Microsoft::UI::Dispatching::DispatcherQueueTimer  lost_face_timer{nullptr};

        explicit Impl(IGameView& v) : view(v)
        {
            dispatcher = winrt::Microsoft::UI::Dispatching::DispatcherQueue::GetForCurrentThread();
            // GameEngine must be constructed on the UI thread.
            // If GetForCurrentThread() returns nullptr, callers used the wrong thread.
            delay_timer     = dispatcher.CreateTimer();
            countdown_timer = dispatcher.CreateTimer();
            lost_face_timer = dispatcher.CreateTimer();
        }

        void Transition(GameState next)
        {
            state = next;
        }

        void DelayOnce(Duration d, std::function<void()> fn)
        {
            delay_timer.Stop();
            delay_timer.Interval(d);
            delay_timer.IsRepeating(false);
            // Replace the previous Tick handler. revoker pattern would be cleaner;
            // for MVP we just keep one shared timer and overwrite.
            static winrt::event_token token{};
            if (token.value)
            {
                delay_timer.Tick(token);
                token = {};
            }
            token = delay_timer.Tick([fn = std::move(fn)](auto&&, auto&&) {
                fn();
            });
            delay_timer.Start();
        }

        void StopAllTimers()
        {
            if (delay_timer)     delay_timer.Stop();
            if (countdown_timer) countdown_timer.Stop();
            if (lost_face_timer) lost_face_timer.Stop();
        }

        // ---- transitions --------------------------------------------------

        void EnterCalibration()
        {
            Transition(GameState::Calibration);
            view.ShowOverlay(L"看着奶龙，先别笑（校准中…）");
            view.RequestSidecarCalibration(true);

            DelayOnce(params.calibration, [this] {
                if (state != GameState::Calibration) return;
                view.RequestSidecarCalibration(false);
                EnterCountdown();
            });
        }

        void EnterCountdown()
        {
            Transition(GameState::Countdown);
            countdown_remaining = 3;
            view.StartCountdown(countdown_remaining);
            view.ShowOverlay(L"3");

            countdown_timer.Stop();
            countdown_timer.Interval(std::chrono::milliseconds{1000});
            countdown_timer.IsRepeating(true);
            static winrt::event_token tok{};
            if (tok.value) { countdown_timer.Tick(tok); tok = {}; }
            tok = countdown_timer.Tick([this](auto&&, auto&&) {
                if (state != GameState::Countdown) { countdown_timer.Stop(); return; }
                --countdown_remaining;
                if (countdown_remaining > 0)
                {
                    view.StartCountdown(countdown_remaining);
                    view.ShowOverlay(countdown_remaining == 2 ? L"2" : L"1");
                }
                else
                {
                    countdown_timer.Stop();
                    view.ClearOverlay();
                    EnterStareLoop();
                }
            });
            countdown_timer.Start();
        }

        void EnterStareLoop()
        {
            Transition(GameState::StareLoop);
            // Arm the hidden deadline.
            std::uniform_int_distribution<long long> dist(
                params.nailong_laugh_min.count(),
                params.nailong_laugh_max.count());
            nailong_laugh_deadline = Clock::now() + Duration{dist(rng)};
            lost_face_since.reset();
            view.BeginStareCycle();
        }

        void DeclareWinner(Winner w)
        {
            last_winner = w;

            // Both outcomes end with nailong laughing. The difference is timing:
            //
            // User wins (nailong laughed first):
            //   t=0     TriggerNailongLaugh → seamless cut at next cycle boundary
            //   t+300ms show "奶龙先笑了！你赢了" overlay, re-enable controls
            //
            // Nailong wins (user laughed first):
            //   t=0     show "你笑了！奶龙赢" overlay (let user read it)
            //   t+1500  TriggerNailongLaugh → nailong also breaks into laughter
            //           re-enable controls (user can reset / start a new round)
            //
            // In both cases the stare cycle keeps running in the background
            // until VideoController.TriggerLaugh() defers the cut to the next
            // Forward→HoldEnd boundary for a visually seamless transition.

            if (w == Winner::User)
            {
                Transition(GameState::NailongLaughTriggered);
                view.TriggerNailongLaugh();
                DelayOnce(Duration{300}, [this, w] {
                    Transition(GameState::Result);
                    view.ShowResult(w);
                });
            }
            else  // Winner::Nailong — user laughed first
            {
                Transition(GameState::UserLaughDetected);
                view.ShowOverlay(L"你笑了！奶龙赢");
                DelayOnce(Duration{1500}, [this, w] {
                    view.TriggerNailongLaugh();
                    Transition(GameState::Result);
                    view.ShowResult(w);
                });
            }
        }

        void DeclareInvalid()
        {
            Transition(GameState::Invalid);
            view.ShowOverlay(L"奶龙看不见你了 — 本局作废");
            DelayOnce(Duration{1500}, [this] {
                Reset();
            });
        }

        void Reset()
        {
            StopAllTimers();
            view.ClearOverlay();
            view.RequestSidecarCalibration(false);  // ensure sidecar is back to idle
            last_winner = Winner::None;
            lost_face_since.reset();
            Transition(GameState::Idle);
        }

        // ---- inputs --------------------------------------------------------

        void OnSample(const SmileSample& s)
        {
            // Track lost-face windows in any "live" state.
            const bool live =
                state == GameState::Countdown ||
                state == GameState::StareLoop;

            if (live)
            {
                if (!s.face_found)
                {
                    if (!lost_face_since) lost_face_since = s.t;
                    auto missing = std::chrono::duration_cast<Duration>(s.t - *lost_face_since);
                    if (missing > params.lost_face_grace)
                    {
                        view.ShowOverlay(L"奶龙看不见你了…");
                    }
                    if (missing > params.lost_face_invalid)
                    {
                        DeclareInvalid();
                        return;
                    }
                }
                else
                {
                    if (lost_face_since)
                    {
                        view.ClearOverlay();
                        lost_face_since.reset();
                    }
                }
            }

            // Smile triggers user-loss only during StareLoop.
            if (state == GameState::StareLoop && s.is_smiling && s.calibrated)
            {
                DeclareWinner(Winner::Nailong);  // user laughed first → nailong wins
            }
        }
    };

    // ---- public surface ----------------------------------------------------

    GameEngine::GameEngine(IGameView& view) : m_impl(std::make_unique<Impl>(view)) {}
    GameEngine::~GameEngine() = default;

    void GameEngine::SetDifficulty(Difficulty d)
    {
        m_impl->difficulty = d;
        m_impl->params = ParamsFor(d);
    }

    void GameEngine::StartChallenge()
    {
        if (m_impl->state != GameState::Idle && m_impl->state != GameState::Result &&
            m_impl->state != GameState::Invalid)
        {
            return;  // already in a round
        }
        m_impl->Transition(GameState::CameraWarmup);
        m_impl->view.ShowOverlay(L"准备摄像头…");
        // GameEngine waits for OnSidecarReady() to be called by the host.
    }

    void GameEngine::Reset()
    {
        m_impl->Reset();
    }

    void GameEngine::OnSmileSample(const SmileSample& s)
    {
        m_impl->OnSample(s);
    }

    void GameEngine::OnStareCycleBoundary()
    {
        if (m_impl->state != GameState::StareLoop) return;
        if (Clock::now() >= m_impl->nailong_laugh_deadline)
        {
            m_impl->DeclareWinner(Winner::User);  // nailong laughed first → user wins
        }
        // else: VideoController will keep cycling on its own.
    }

    void GameEngine::OnSidecarReady()
    {
        if (m_impl->state != GameState::CameraWarmup) return;
        m_impl->EnterCalibration();
    }

    void GameEngine::OnSidecarLost()
    {
        m_impl->StopAllTimers();
        m_impl->view.ShowOverlay(L"摄像头检测进程已断开");
        m_impl->Transition(GameState::Invalid);
    }

    GameState GameEngine::State() const noexcept     { return m_impl->state; }
    Winner    GameEngine::LastWinner() const noexcept { return m_impl->last_winner; }
}

File: desktop_pet/game_engine.py
"""
game_engine.py — 9-state FSM for the don't-laugh challenge.

Faithful port of Core/GameEngine.h/cpp + Core/GameState.h.
All methods must be called from the GUI thread.
"""

from __future__ import annotations

import enum
import random
import time
from dataclasses import dataclass
from typing import Protocol

from PySide6.QtCore import QObject, QTimer

from config import DifficultyParams


# ── enums (mirror GameState.h) ─────────────────────────────────────


class GameState(enum.IntEnum):
    Idle = 0
    CameraWarmup = 1
    Calibration = 2
    Countdown = 3
    StareLoop = 4
    UserLaughDetected = 5
    NailongLaughTriggered = 6
    Result = 7
    Invalid = 8


class Winner(enum.IntEnum):
    NoOne = 0
    User = 1       # nailong laughed first → user wins
    Nailong = 2    # user laughed first → nailong wins


class Difficulty(enum.IntEnum):
    Easy = 0
    Normal = 1
    Hard = 2


# ── data ───────────────────────────────────────────────────────────


@dataclass
class SmileSample:
    t: float              # monotonic time (seconds)
    face_found: bool
    smile_score: float    # baseline-subtracted, 0..1
    is_smiling: bool      # sliding-window confirmed
    calibrated: bool


# ── view protocol ──────────────────────────────────────────────────


class IGameView(Protocol):
    def show_overlay(self, text: str) -> None: ...
    def clear_overlay(self) -> None: ...
    def start_countdown(self, seconds: int) -> None: ...
    def begin_stare_cycle(self) -> None: ...
    def trigger_nailong_laugh(self) -> None: ...
    def show_result(self, winner: Winner) -> None: ...
    def request_calibration(self, start: bool) -> None: ...


# ── engine ─────────────────────────────────────────────────────────


class GameEngine(QObject):
    """Central game state machine — port of Core/GameEngine."""

    def __init__(self, view: IGameView, parent: QObject | None = None):
        super().__init__(parent)
        self._view = view
        self._state = GameState.Idle
        self._difficulty = Difficulty.Normal
        self._params = DifficultyParams()
        self._last_winner = Winner.NoOne

        self._nailong_laugh_deadline: float = 0.0
        self._lost_face_since: float | None = None
        self._countdown_remaining: int = 0

        # Timers (mirror the 3 DispatcherQueueTimers in C++)
        self._delay_timer = QTimer(self)
        self._delay_timer.setSingleShot(True)

        self._countdown_timer = QTimer(self)
        self._countdown_timer.setInterval(1000)

    # ── properties ─────────────────────────────────────────────

    @property
    def state(self) -> GameState:
        return self._state

    @property
    def last_winner(self) -> Winner:
        return self._last_winner

    # ── UI inputs ──────────────────────────────────────────────

    def set_difficulty(self, d: Difficulty, params: DifficultyParams) -> None:
        self._difficulty = d
        self._params = params

    def start_challenge(self) -> None:
        """Idle/Result/Invalid → CameraWarmup."""
        if self._state not in (GameState.Idle, GameState.Result, GameState.Invalid):
            return
        self._state = GameState.CameraWarmup
        self._view.show_overlay("准备摄像头…")

    def reset(self) -> None:
        """Any state → Idle."""
        self._stop_all_timers()
        self._view.clear_overlay()
        self._view.request_calibration(False)
        self._last_winner = Winner.NoOne
        self._lost_face_since = None
        self._state = GameState.Idle

    # ── detector inputs ────────────────────────────────────────

    def on_detector_ready(self) -> None:
        """Called when first valid face is detected."""
        if self._state != GameState.CameraWarmup:
            return
        self._enter_calibration()

    def on_detector_lost(self, reason: str = "") -> None:
        """Called when camera/model fails."""
        self._stop_all_timers()
        self._view.show_overlay("摄像头检测进程已断开")
        self._state = GameState.Invalid

    def on_smile_sample(self, s: SmileSample) -> None:
        """Process a smile detection sample."""
        # Track lost-face in any "live" state
        live = self._state in (GameState.Countdown, GameState.StareLoop)

        if live:
            if not s.face_found:
                if self._lost_face_since is None:
                    self._lost_face_since = s.t
                missing_ms = (s.t - self._lost_face_since) * 1000
                if missing_ms > self._params.lost_face_grace_ms:
                    self._view.show_overlay("奶龙看不见你了…")
                if missing_ms > self._params.lost_face_invalid_ms:
                    self._declare_invalid()
                    return
            else:
                if self._lost_face_since is not None:
                    self._view.clear_overlay()
                    self._lost_face_since = None

        # Smile triggers user-loss only during StareLoop
        if (self._state == GameState.StareLoop
                and s.is_smiling and s.calibrated):
            self._declare_winner(Winner.Nailong)  # user laughed → nailong wins

    # ── video controller input ─────────────────────────────────

    def on_stare_cycle_boundary(self) -> None:
        """Called at end of each forward-stare phase."""
        if self._state != GameState.StareLoop:
            return
        if time.monotonic() >= self._nailong_laugh_deadline:
            self._declare_winner(Winner.User)  # nailong laughed first → user wins

    # ── internal transitions ───────────────────────────────────

    def _stop_all_timers(self) -> None:
        self._delay_timer.stop()
        self._countdown_timer.stop()

    def _delay_once(self, ms: int, fn) -> None:
        """Fire fn after ms milliseconds (single-shot)."""
        self._delay_timer.stop()
        try:
            self._delay_timer.timeout.disconnect()
        except TypeError:
            pass
        self._delay_timer.setInterval(ms)
        self._delay_timer.timeout.connect(fn)
        self._delay_timer.start()

    def _enter_calibration(self) -> None:
        self._state = GameState.Calibration
        self._view.show_overlay("看着奶龙，先别笑（校准中…）")
        self._view.request_calibration(True)

        def on_done():
            if self._state != GameState.Calibration:
                return
            self._view.request_calibration(False)
            self._enter_countdown()

        self._delay_once(self._params.calibration_ms, on_done)

    def _enter_countdown(self) -> None:
        self._state = GameState.Countdown
        self._countdown_remaining = 3
        self._view.start_countdown(self._countdown_remaining)
        self._view.show_overlay("3")

        self._countdown_timer.stop()
        try:
            self._countdown_timer.timeout.disconnect()
        except TypeError:
            pass

        def on_tick():
            if self._state != GameState.Countdown:
                self._countdown_timer.stop()
                return
            self._countdown_remaining -= 1
            if self._countdown_remaining > 0:
                self._view.start_countdown(self._countdown_remaining)
                self._view.show_overlay(str(self._countdown_remaining))
            else:
                self._countdown_timer.stop()
                self._view.clear_overlay()
                self._enter_stare_loop()

        self._countdown_timer.timeout.connect(on_tick)
        self._countdown_timer.start()

    def _enter_stare_loop(self) -> None:
        self._state = GameState.StareLoop
        delay_ms = random.randint(
            self._params.nailong_laugh_min_ms,
            self._params.nailong_laugh_max_ms,
        )
        self._nailong_laugh_deadline = time.monotonic() + delay_ms / 1000.0
        self._lost_face_since = None
        self._view.begin_stare_cycle()

    def _declare_winner(self, w: Winner) -> None:
        self._last_winner = w

        if w == Winner.User:
            # Nailong laughed first — trigger laugh, then show result after 300ms
            self._state = GameState.NailongLaughTriggered
            self._view.trigger_nailong_laugh()

            def show():
                self._state = GameState.Result
                self._view.show_result(w)

            self._delay_once(300, show)
        else:
            # User laughed first — show overlay, then trigger nailong laugh after 500ms
            self._state = GameState.UserLaughDetected
            self._view.show_overlay("你笑了！奶龙赢")

            def laugh_and_show():
                self._view.trigger_nailong_laugh()
                self._state = GameState.Result
                self._view.show_result(w)

            self._delay_once(500, laugh_and_show)

    def _declare_invalid(self) -> None:
        self._state = GameState.Invalid
        self._view.show_overlay("奶龙看不见你了 — 本局作废")
        self._delay_once(1500, self.reset)
